### ISSUE TITLE
Exception semantics

### ISSUE THREAD

**Issue by dmoisset** [gvanrossum/patma#111]
Something that is not 100% clear to me is what happens with exceptions raised while matching. The only description I found is [here](https://www.python.org/dev/peps/pep-0622/#guards) «If evaluating a guard raises an exception, it is propagated onwards rather than fail the case clause». So exceptions in a guard are propagated outside of the match statement... the way it's written seem to imply that if the exception is NOT in a guard, they cause a fail to match (i.e. the matching keeps attempting with the next case clause?)

I'm thinking that all of these could happen:
 * A `__match__` method could raise; that sounds like something that should bubble up.
 * An error on `__eq__` while comparing with a literal should bubble up.
 * A name lookup error (`NameError` and `AttributeError` in either a constant or the class name of a class pattern) should bubble up.
 * Failure in something like `__getitem__` or perhaps `__len__` should bubble up (I'm less sure about this one, do I care if I was mathing with `_`, for example if I have some sort of sparse sequence?)
 * Using a class pattern that accesses an attribute which is a descriptor, and the descriptors' `__get__` throws a weird exception. This is the one I don't know about. `AttributeError` already has special treatment here (mismatch rather than bubble up) if the missing attribute's name is in `__match_args__`, should other exceptions be treated the same?

To summarise: is the desired behaviour "by default, any exception raised by special methods called by matching bubble up"? what are exactly the situations where an exception doesn't bubble up? (besides the `AttributeError` case)

And finally: if exception behaviour is being clarified, some explanation on why the special case(s) are useful would be helpful.

**Comment by gvanrossum** [comment]
Exceptions should never be caught by the match machinery, except when used to check for the presence of an attribute. But that should probably be done using getattr(x, name, marker) to avoid races.

**Comment by gvanrossum** [comment]
Sorry, on the last bullet: I thought we got rid of the special case. Maybe there’s an editing mistake?

**Comment by gvanrossum** [comment]
@brandtbucher  Specifically, the PEP still states that if an attribute is missing on the proxy and `__match_args__` is present and the attribute is not in the latter, the match raises `ImpossibleMatchError`. I think this is undesirable (and I thought we had decided not to do that, but I can't recall where). The reason I think it is undesirable is that it makes it impossible to have attributes that are optional (may be missing) and cannot be passed positionally. (And yes, it's probably not great design to have attributes that may or may not be set, but I've seen this pattern enough that I don't want to have to tell people they can't use it.) It also makes the description of `__match__` more complicated.

**Comment by gvanrossum** [comment]
> * Failure in something like `__getitem__` or perhaps `__len__` should bubble up (I'm less sure about this one, do I care if I was matching with `_`, for example if I have some sort of sparse sequence?)

The target for a sequence patterns can only match when `isinstance(target, collections.abc.Sequence)` (and similar for mapping patterns and `Mapping`). Those implement ‘should implement `__len__` and if something fails to return a length it should bubble up (just like `__eq__`). The matching machinery shouldn't attempt `__getitem__` for a wildcard, but if we have e.g.
```
match mystery_sequence:
  case [1, x, y]: ...
  case [_, _, _]: ...
```
and the sequence raises on one of the three `__getitem__` accesses needed to implement the first case, I think it should bubble up.

> * Using a class pattern that accesses an attribute which is a descriptor, and the descriptors' `__get__` throws a weird exception. This is the one I don't know about.

Huh, so you're thinking of an example like this:
```
class deco:
    def __init__(self, func): pass
    def __get__(self, obj, cls): raise RuntimeError
class C:
    @deco
    def foo(self): ...
match C():
    case object(foo=_): print("A")
    case _: print("B")
```
@brandtbucher The current implementation prints "B", but I think it should bubble up the `RuntimeError`.

> `AttributeError` already has special treatment here (mismatch rather than bubble up) if the missing attribute's name is in `__match_args__`, should other exceptions be treated the same?

I think `AttributeError` (for an attribute of the target probed by a class pattern) should always fail instead of raising. See https://github.com/python/peps/pull/1477 which I just created to make it so in the PEP text. I don't think any other exceptions should be turned into failure to match (it would be bad if this could suppress e.g. `^C` or `MemoryError`).

**Comment by brandtbucher** [comment]
> Specifically, the PEP still states that if an attribute is missing on the proxy and `__match_args__` is present and the attribute is not in the latter, the match raises `ImpossibleMatchError`. I think this is undesirable (and I thought we had decided not to do that, but I can't recall where).

I don't remember this reversal, but I'm fine with it. As you say, it makes things simpler. I also think its only motivation was to catch typos, but one huge advantage of our simpler match protocol is that the default `__match__` and custom `__match_args__` play very nicely with static analysis tools, so I doubt typos will be much of an issue in practice.

> The current implementation prints "B", but I think it should bubble up the RuntimeError.

Yeah, I have an outstanding TODO to consider only catching `AttributeError` there (the interaction with the `__match_args__` lookup wasn't obvious to me). But now I have my answer!

I'll make both of these changes today.

**Comment by brandtbucher** [comment]
The cases where we raise an `ImpossibleMatchError` are becoming fewer and fewer (there are only four now). I wonder if we should just ditch it and use existing exceptions? Here are all of the places we raise.

Could be `TypeError`:
- `type %s cannot be matched` (types setting `__match__ = None`)
- `%s() takes at most %d positional sub-patterns (%d given)`
- `%s() got multiple sub-patterns for attribute %R`

Could be `ValueError`:
- `mapping pattern checks duplicate key (%R)`

**Comment by gvanrossum** [comment]
Okay, so I landed my simplification of `__match__`(https://github.com/python/peps/pull/1477).

Regarding `ImpossibleMatchError` I'm fine with tossing it. For the Walrus PEP (PEP 572) we originally planned a separate exception but never implemented it, and eventually we dropped it from the PEP. Nobody regrets that.

**Comment by gvanrossum** [comment]
(We dropped `ImpossibleMatchError`: 5aec654f9e056ca9990e6b191b6b5f76f64eeec2.)

**Comment by dmoisset** [comment]
I added a preview PR (in my local fork) of what I'd like to add for review. If you think this helps, I can submit the PR against the peps repo.

**Comment by dmoisset** [comment]
Ok, I've pushed this to the PEPs repo. (I've signed the CLA already but it's being processed by the PSF I assume).

**Comment by gvanrossum** [comment]
This landed as 127fc3701e653d12a5850b2eed524357b64da16b
